cpp/src/hash/hashing.cu

```diff
@@ -607,6 +611,47 @@ std::pair<std::unique_ptr<table>, std::vector<size_type>> hash_partition_table(
   }
 }
 
+/**
+ * @brief Finalize MD5 hash including converstion to hex string.
+ */
+void __device__ finalize_md5_hash(detail::md5_intermediate_data* hash_state,
+                                  char* result_location,
+                                  const detail::md5_hash_constants_type* hash_constants,
+                                  const detail::md5_shift_constants_type* shift_constants,
+                                  const detail::hex_to_char_mapping_type* hex_char_map)
+{
+  auto const full_length = (static_cast<uint64_t>(hash_state->message_length)) << 3;
+  thrust::fill_n(thrust::seq, hash_state->buffer + hash_state->buffer_length, 1, 0x80);
+
+  if (hash_state->buffer_length < 56) {
+    thrust::fill_n(thrust::seq,
+                   hash_state->buffer + hash_state->buffer_length + 1,
+                   (55 - hash_state->buffer_length),
+                   0x00);
+  } else {
+    thrust::fill_n(thrust::seq,
+                   hash_state->buffer + hash_state->buffer_length + 1,
+                   (64 - hash_state->buffer_length),
+                   0x00);
+    detail::md5_hash_step(hash_state, hash_constants, shift_constants);
+
+    thrust::fill_n(thrust::seq, hash_state->buffer, 56, 0x00);
+  }
+
+  thrust::copy_n(
+    thrust::seq, reinterpret_cast<uint8_t const*>(&full_length), 8, hash_state->buffer + 56);
+  detail::md5_hash_step(hash_state, hash_constants, shift_constants);
+
+  u_char final_hash[32];
+  uint8_t* hash_result = reinterpret_cast<uint8_t*>(hash_state->hash_value);
+  for (int i = 0; i < 16; i++) {
+    final_hash[i * 2]     = hex_char_map[(hash_result[i] >> 4) & 0xf];
+    final_hash[i * 2 + 1] = hex_char_map[hash_result[i] & 0xf];
```

Comment:
computing this should be faster than accessing an array (in this case global, cached). 
Try
`auto hex_char_map = [](auto i) { return (i<0xa ? '0'+i : 'a'-0xa+i); }`